plugins/logs/plugin.go

```diff
@@ -711,7 +713,20 @@ func (p *Plugin) maskEvent(ctx context.Context, txn storage.Transaction, event *
 		return err
 	}
 
+	resultCopied := false
 	for _, mRule := range mRules {
+		// result must be deep copied if there are any mask rules
+		// targeting it, to avoid modifying the result sent
+		// to the consumer
+		if mRule.escapedParts[0] == "result" && !resultCopied {
+			resultCopy, err := deepCopy(*event.Result)
+			if err != nil {
+				p.logError("failed to deep copy result for masking, rule skipped: %s: %s", mRule.String(), err.Error())
+				continue
+			}
+			event.Result = &resultCopy
+			resultCopied = true
+		}
```

Comment:
Nit: let's do the deep copy inside the mask implementation rather than outside: https://github.com/open-policy-agent/opa/blob/master/plugins/logs/mask.go#L145

Reply:
`mRule.Mask` is iterated on though, and we only want to deep copy Result once, right?  We have to keep the state tracking whether or not Result had been modified somewhere, which is why I did it outside the loop and in plugin.go. If we do it in `mRule.Mask` the only spot I can see to store that state would be on event itself, is that ok?  If that works, or if there's another option I'm not seeing, I'm happy to make the change.

Reply:
> mRule.Mask is iterated on though, and we only want to deep copy Result once, right?

Ah, good catch. I haven't looked at the masking in a while.

> If that works, or if there's another option I'm not seeing, I'm happy to make the change.

I think what I would do is just modify `resultValueToMaskRules` to return a struct (e.g., `maskRuleSet`) rather than a raw slice. Then we could put the iteration into the struct implementation and do the deep copy there. That way the details of the masking are handled in one place as opposed to being split across the plugin and the mask.go file.

Reply:
Gotcha, that sounds reasonable. I'll take a crack at that soon.  Thanks for the feedback!